pkg/apis/apps/v1alpha1/uniteddeployment_types.go

```diff
@@ -0,0 +1,229 @@
+	// If unspecified, defaults to 20
+	// +optional
+	RevisionHistoryLimit *int32 `json:"historyLimit,omitempty"`
+}
+
+// SubsetTemplate defines the subset under the UnitedDeployment
+type SubsetTemplate struct {
+	// Indicate the type of the underlying workload
+	// If unspecified, defaults to ReplicaSet
+	// +optional
+	SubsetType *metav1.TypeMeta `json:"subsetType,omitempty"`
+
+	// StatefulSet template
+	StatefulSet *StatefulSetTemplateSpec `json:"statefulSet,omitempty"`
+
+	// ReplicaSet template
+	ReplicaSet *ReplicaSetTemplateSpec `json:"replicaSet,omitempty"`
+}
+
+// StatefulSetTemplateSpec defines the subset template of StatefulSet
+type StatefulSetTemplateSpec struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              appsv1.StatefulSetSpec `json:"spec,omitempty"`
+}
+
+// ReplicaSetTemplateSpec defines the subset template of ReplicasSet
+type ReplicaSetTemplateSpec struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              appsv1.ReplicaSetSpec `json:"spec,omitempty"`
+}
+
+// UnitedDeploymentUpdateStrategy defines the update strategy of UnitedDeployment
+type UnitedDeploymentUpdateStrategy struct {
+	// Pause the rollout of all subset worloads
+	// +optional
+	Pause bool `json:"pause,omitempty"`
+
+	// Indicate the partition of each subset
+	Partitions map[string]int32 `json:"partitions,omitempty"`
+}
+
+// Topology defines the spread detail of each subset under UnitedDeployment
+type Topology struct {
+	// Contains the details of each subset
+	Subsets []Subset `json:"subsets,omitempty"`
+}
+
+// Subset defines the detail of a subset
+type Subset struct {
+	// Indicate the name of this subset, which will be used to generate
+	// subset workload name in the format '<deployment-name>-<subset-name>'
+	Name string `json:"name,omitempty"`
+
+	// Indicate the node select strategy to form the subset
+	NodeSelector corev1.NodeSelector `json:"nodeSelector,omitempty"`
+
+	// Indicate the number of pod replicas of this subset.
+	// If nil, the number of replicas in this subset is determined by controller.
+	// +optional
+	Replicas *intstr.IntOrString `json:"replicas,omitempty"`
+}
+
+// UnitedDeploymentStatus defines the observed state of UnitedDeployment
+type UnitedDeploymentStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	// observedGeneration is the most recent generation observed for this InPlaceSet. It corresponds to the
+	// InPlaceSet's generation, which is updated on mutation by the API Server.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// selector is a label query over pods that should match the replica count.
+	// It must match the template's labels.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	Selector *metav1.LabelSelector `json:"selector"`
```

Comment:
Why is `Selector` needed?

Reply:
This is interesting to me, too

Reply:
In case of ReplicaSet, UnitedCafeDeployment will create a new ReplicaSet to rolling update, so we need `selector` to choose the set of subsets.

The `selector` and `ownerReference` work together to build the relationship between unitedDeployment and its subsets like statefulSet or replicaSet.
And the `subset.name` is used to figure out which subset a statefulSet or replicaSet belongs to.

Recently, we use the `subset.name` to generate the name of a subse in format `<deployment-name>-<subset-name>`, or its prefix in format `<deployment-name>-<subset-name>-` in case of ReplicaSet. It seems not to be consistent. How about we just put the `subset.name` in the label of each subset. And keep the subset name in format `<deployment-name>-xxxx` which  just like k8s deployment and replicaSet do.